### PR TITLE
WIP: Poison update

### PR DESCRIPTION
**Pull request by davidslater**

### CODE REVIEW

File: armory/scenarios/poison.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+    def _evaluate(self) -> dict:
+        """
+        Evaluate a config file for classification robustness against attack.
+        """
+        self.set_random_seed()
+        self.load_model()
+        self.load_defense()
+        self.load_train_dataset()
+        self.poison_dataset()
```
Comment (ng390) [inline armory/scenarios/poison.py:328]: I think ```self.load_attack()``` needs to be called before this line -- the ```poison_dataset``` method references self.use_poison, which is initialized in ```self.load_attack()```

File: armory/scenarios/poisoning_gtsrb_clbd.py [hunk 2]
```diff
@@ -1,325 +1,61 @@
...
-        img_eq = np.array(img_eq.resize([img_size, img_size])) / quantization
+        if return_index:
+            logger.warning("return_index not currently implemeneted. Returning []")
+            x_poison, y_poison, np.array([])
+        return x_poison, y_poison
 
-        img_out.append(img_eq)
 
-    return np.array(img_out, dtype=np.float32)
+class CleanGTSRB(GTSRB):
```
Comment (ng390) [inline armory/scenarios/poisoning_gtsrb_clbd.py:26]: The name reference in the configs needs to update to reflect the class name change from GTSRB_CLBD.

File: armory/scenarios/poison.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+                attack,
+                self.source_class,
+                self.target_class,
+                fraction=adhoc_config["fraction_poisoned"],
+            )
+            self.test_poisoner = self.poisoner
+
+    def poison_dataset(self):
+        if self.use_poison:
+            self.x_poison, self.y_poison, self.poison_index = self.poisoner.poison(
```
Comment (ng390) [inline armory/scenarios/poison.py:171]:
```suggestion
            self.x_poison, self.y_poison, self.poison_index = self.poisoner.poison_dataset(
```

File: armory/scenarios/poison.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+            raise ValueError("Sizes of x and y do not match")
+        poison_x, poison_y = list(x), list(y)
+        poison_index = self.get_poison_index(y, fraction=fraction)
+        for i in poison_index:
+            poison_x_i, poison_y[i] = self.attack.poison(x[i], [self.target_class])
+            poison_x[i] = np.asarray(poison_x_i, dtype=x[i].dtype)
+        poison_x, poison_y = np.array(poison_x), np.array(poison_y)
+
+        if return_index:
+            poison_x, poison_y, poison_index
```
Comment (ng390) [inline armory/scenarios/poison.py:75]:
```suggestion
            return poison_x, poison_y, poison_index
```

File: armory/scenarios/poison.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+                )
+
+    def test(self):
+        if self.use_poison:
+            logger.info("Testing on poisoned test data")
+            test_data = config_loading.load_dataset(
+                self.config["dataset"], split=self.test_split, **self.dataset_kwargs
+            )
+            for x, y in tqdm(test_data, desc="Testing"):
+                x, _ = self.test_poisoner.poison(x, y, fraction=1.0)
```
Comment (ng390) [inline armory/scenarios/poison.py:286]:
```suggestion
                x, _ = self.test_poisoner.poison_dataset(x, y, fraction=1.0)
```

File: armory/scenarios/poisoning_gtsrb_clbd.py [hunk 6]
```diff
@@ -1,325 +1,61 @@
...
-        right = center[0] + min_side // 2
-        bottom = center[1] + min_side // 2
+    def poison_dataset(self, x, y, return_index=False):
+        x_poison, y_poison = self.attack.poison(x, y)
 
-        img_eq = img_eq.crop((left, top, right, bottom))
-        img_eq = np.array(img_eq.resize([img_size, img_size])) / quantization
+        if return_index:
+            logger.warning("return_index not currently implemeneted. Returning []")
+            x_poison, y_poison, np.array([])
```
Comment (ng390) [inline armory/scenarios/poisoning_gtsrb_clbd.py:22]:
```suggestion
            return x_poison, y_poison, np.array([])
```

File: armory/scenarios/poisoning_gtsrb_clbd.py [hunk 7]
```diff
@@ -1,325 +1,61 @@
...
+class CleanDatasetPoisoner:
+    def __init__(self, attack):
+        self.attack = attack
 
-        left = center[0] - min_side // 2
-        top = center[1] - min_side // 2
-        right = center[0] + min_side // 2
-        bottom = center[1] + min_side // 2
+    def poison_dataset(self, x, y, return_index=False):
+        x_poison, y_poison = self.attack.poison(x, y)
```
Comment (ng390) [inline armory/scenarios/poisoning_gtsrb_clbd.py:18]: The main ART attack used in practice here (```PoisoningAttackCleanLabelBackdoor```) expects y in a one-hot format, and returns one-hot labels. It throws a numpy.AxisError when the y input is sparse. To avoid cluttering the scenario code, I guess we could modify the attack.

File: armory/scenarios/poisoning_gtsrb_clbd.py (hunk 2 above)
Comment (davidslater) [reply armory/scenarios/poisoning_gtsrb_clbd.py:26]: I'll change it back to the original name.

File: armory/scenarios/poisoning_gtsrb_clbd.py (hunk 6 above)
Comment (davidslater) [reply armory/scenarios/poisoning_gtsrb_clbd.py:22]: As an aside, do you know how to get which indices are poisoned in the clean label example? It would be helpful for quantifying filtering defenses?

File: armory/scenarios/poisoning_gtsrb_clbd.py (hunk 7 above)
Comment (davidslater) [reply armory/scenarios/poisoning_gtsrb_clbd.py:18]: I had removed some lines for `to_categorical` and back - I'll add them back in.

File: armory/scenarios/poisoning_gtsrb_clbd.py (hunk 6 above)
Comment (ng390) [reply armory/scenarios/poisoning_gtsrb_clbd.py:22]: Unfortunately, since the attack doesn't return that, one has to do a pixel-level comparison with the original image (since only poisoned images are changed).

Comment (davidslater) [reply armory/scenarios/poisoning_gtsrb_clbd.py]: okay, I think I made code for that

**Comment by davidslater** [comment]
I think I'm ready for another review. @ng390

File: armory/scenarios/poison.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+            )
+            self.test_poisoner = self.poisoner
+
+    def poison_dataset(self):
+        if self.use_poison:
+            (
+                self.x_poison,
+                self.y_poison,
+                self.poison_index,
+            ) = self.poisoner.poison_datset(
```
Comment (ng390) [inline armory/scenarios/poison.py:167]:
```suggestion
            ) = self.poisoner.poison_dataset(
```

**Comment by davidslater** [comment]
Replaced by #1117
